sklearn/manifold/tests/test_t_sne.py

```diff
@@ -718,25 +720,33 @@ def check_uniform_grid(method, seeds=[0, 1, 2], n_iter=1000):
     """Make sure that TSNE can approximately recover a uniform 2D grid"""
     for seed in seeds:
         tsne = TSNE(n_components=2, init='random', random_state=seed,
-                    perplexity=10, n_iter=n_iter, method=method)
+                    perplexity=20, n_iter=n_iter, method=method)
         Y = tsne.fit_transform(X_2d_grid)
 
-        # Ensure that the convergence criterion has been triggered
-        assert tsne.n_iter_ < n_iter
+        try_name = "{}_{}".format(method, seed)
+        try:
+            assert_uniform_grid(Y, try_name)
+        except AssertionError:
+            # If the test fails a first time, re-run with init=Y to see if
+            # this was caused by a bad initialization
+            tsne.init = Y
```

Comment:
Instead of this, what do you think, @tommoral, of just having a counter of how many times the uniform grid was recovered, and checking that, say, 2 seeds of 3 were good.

Reply:
I would say that 3 tries are not enough for this 2 out of 3 approach. However, the test is already slow.

I think the proposed solution is more robust. It asserts that we are able to recover a 2d grid from a random initialization using t-SNE. The retry is just caused by the fact that you can have some bad initialization configuration but does not break the fact that we are converging to the right solution with an optimization schedule good enough.